expression/builtin_cast_vec.go

```diff
@@ -170,11 +170,35 @@ func (b *builtinCastDecimalAsStringSig) vecEvalString(input *chunk.Chunk, result
 }
 
 func (b *builtinCastTimeAsDecimalSig) vectorized() bool {
-	return false
+	return true
 }
 
 func (b *builtinCastTimeAsDecimalSig) vecEvalDecimal(input *chunk.Chunk, result *chunk.Column) error {
-	return errors.Errorf("not implemented")
+	n := input.NumRows()
+	buf, err := b.bufAllocator.get(types.ETDatetime, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf)
+	if err := b.args[0].VecEvalTime(b.ctx, input, buf); err != nil {
+		return err
+	}
+	result.ResizeDecimal(n, false)
+	result.MergeNulls(buf)
+	times := buf.Times()
+	decs := result.Decimals()
+	sc := b.ctx.GetSessionVars().StmtCtx
+	for i := 0; i < n; i++ {
+		if result.IsNull(i) {
+			continue
+		}
+		decimalValue, err := types.ProduceDecWithSpecifiedTp(times[i].ToNumber(), b.tp, sc)
+		if err != nil {
+			return err
+		}
+		decs[i] = *decimalValue
```

Comment:
It does extra data copy comparing to `evalDecimal`.
We can add a method for time naming ``times[i].ToDecimail(&desc[i])` to prevent `ToNumber` to allocate memory itself
So your code can be modified like
```go
times[i].ToDecimail(&desc[i])
_, err := types.ProduceDecWithSpecifiedTp(&desc[i], b.tp, sc)
```

Reply:
@polyrabbit 
Would you like to create another pull request to address this?
It may improve the benchmark result.

Reply:
OK, I'll send another PR.

Reply:
Done, please see PR #12439